_posts/2019-04-24-istio-service-mesh-s2s.md

```diff
@@ -0,0 +1,431 @@
+kubectl label namespace with-istio istio-injection=enabled
+```
+
+### Install CouchDB
+
+Next, the database is installed.
+This looks like a normal stateful set for a CouchDB database.
+There are some important changes.
+
+First, a specific service account is created for CouchDB. 
+This is needed as Istio will use the service accounts in Kubernetes as its identities.
+The service account is linked to the podspec in the stateful set definition. 
+This way, it can be used by the Istio proxy later on.
+
+Secondly, the probes have been adapted to work in Istio.
+Since Istio intercepts all traffic in the pod, it will also intercept requests from the Kube API to the service.
+Since the demo setup requires mTLS to be used, the probes would fail because the Kube API doesn't use mTLS.
+Instead of using changing the probes, Istio now has the option to rewrite the probes during the automatic proxy injection. 
+More information on the probes can be found in the [Istio Docs](https://istio.io/help/ops/setup/app-health-check/#probe-rewrite){:target="_blank" rel="noopener noreferrer"}.
+
+Note that no Istio specific configuration is required in the service manifests.
+This is possible because the demo profile automatically enables the sidecar injection for all namespaces.
+The automatic sidecar injector will inject the Envoy sidecar into all pods.
+
+```yaml
+---
+# Source: couchdb/templates/serviceaccount.yaml
+apiVersion: v1
+kind: ServiceAccount
+metadata:
+  name: couchdb
+---
+# Source: couchdb/templates/service.yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: couchdb
+  labels:
+   ...
+spec:
+  type: ClusterIP
+  ports:
+    - port: 80
+      targetPort: http
+      protocol: TCP
+      name: http
+  selector:
+    app.kubernetes.io/name: couchdb
+    app.kubernetes.io/instance: couchdb
+---
+# Source: couchdb/templates/statefulset.yaml
+apiVersion: apps/v1
+kind: StatefulSet
+metadata:
+  name: couchdb
+  labels:
+    ...
+spec:
+  replicas: 1
+  serviceName: couchdb
+  selector:
+    matchLabels:
+      app.kubernetes.io/name: couchdb
+      app.kubernetes.io/instance: couchdb
+  template:
+    metadata:
+      labels:
+        app.kubernetes.io/name: couchdb
+        app.kubernetes.io/instance: couchdb
+    spec:
+      serviceAccountName: couchdb
+      containers:
+        - name: couchdb
+          image: "couchdb:2.3.0"
+          imagePullPolicy: IfNotPresent
+          ports:
+            - name: http
+              containerPort: 5984
+              protocol: TCP
+          livenessProbe:
+            exec:
+              command:
+              - curl
+              - http://localhost:5984/_up
+          readinessProbe:
+            exec:
+              command:
+              - curl
+              - http://localhost:5984/_up
+          resources:
+            ...
+```
+
+### Install the service: test-app
+
+A small NodeJS application was created for this demo.
+It exposes a HTTP GET endpoint which connects to the CouchDB database.
```

Comment:
It exposes *an* HTTP GET endpoint which connects to the CouchDB database.

Reply:
Updated